internal/releaseapi/client.go

```diff
@@ -87,48 +91,101 @@ func (c *Client) ListReleases() (ReleaseIndex, error) {
 
 func (c *Client) DownloadRelease(r Release, os, arch string) (string, error) {
 	var matchingBuild Build
-
+	var checkSha256Sum string
 	for _, build := range r.Builds {
 		if build.OS == os && build.Arch == arch {
 			matchingBuild = build
 			break
 		}
 	}
 
+	checkSums, err := c.getReleaseCheckSums(r)
+	if err != nil {
+		return "", errors.Wrap(err, "could not download checksum file")
+	}
+
+	for _, line := range strings.Split(checkSums, "\n") {
+		checksum := strings.Split(line, "  ")
+		if checksum[1] == matchingBuild.zipFileName() {
+			checkSha256Sum = checksum[0]
+			break
+		}
+	}
+
 	if matchingBuild.URL == "" {
 		return "", errors.Errorf(
 			"could not find matching build for OS '%s' and arch '%s'", os, arch,
 		)
 	}
 
-	return c.downloadBuild(matchingBuild)
+	build, sha256sum, err := c.downloadBuild(matchingBuild)
+	if sha256sum != nil { // new download
+		if checkSha256Sum != hex.EncodeToString(sha256sum) {
+			return "", errors.Errorf(
+				"checksum for %s should be %s, got %s", matchingBuild.URL, checkSha256Sum, hex.EncodeToString(sha256sum),
+			)
+		} else {
+			log.Printf("check sum match\n")
+		}
+	}
+
+	return build, err
 }
 
-func (c *Client) downloadBuild(build Build) (string, error) {
+func (c *Client) getReleaseCheckSums(release Release) (string, error) {
+	shaSumFile, _, err := c.downloadReleaseArchive(release.ShaSumsURL())
```

Comment:
Another thought - that would mean we wouldn't need to write the shasum file to disk, since it'd get cached by the HTTP caching library anyways, iiuc. `downloadReleaseArchive` turns off caching because it's downloading the zip file, so we're not getting that benefit as it is written now.